pkg/error_code/error_code.go

```diff
@@ -328,6 +347,73 @@ func NewNotFoundErr(err error) ErrorCode {
 var _ ErrorCode = (*notFoundErr)(nil)     // assert implements interface
 var _ HasClientData = (*notFoundErr)(nil) // assert implements interface
 
+// HasOperation defines the operation that occurred during an error
+// GetOperation is defined, but generally Operation() should be used.
+// Operation() will check for a HasOperation definition or an Operation field.
+// You don't need to define this interface if your struct has an Operation field.
+type HasOperation interface {
+	GetOperation() string
+}
+
+// Operation will return an operation string if it exists.
+// It looks for either the HasOperation interface or a struct field "Operation".
+// Otherwise it will return the zero value (empty) string.
+func Operation(v interface{}) string {
+	var operation string
+	if hasOp, ok := v.(HasOperation); ok {
+		operation = hasOp.GetOperation()
+	} else {
+		operation = StructFieldOperationValue(v)
+	}
+	return operation
+}
+
+// StructFieldOperationValue looks for a struct field "Operation" of type String.
+// Return that or the empty string.
+func StructFieldOperationValue(i interface{}) string {
```

Comment:
IMO, I don't like using dynamic refection in the project, can we have another way to ignore this? do we really need to care the operation field in a struct? can we implement the HasOperation for these structs?

Reply:
Yes, one can implement `HasOperation` and not use this, so from that sense it is optional. However, I thought the overhead of implementing the interface is a lot compared to just having an `Operation` field on a struct, to the point that I might hesitate to do that. There is just one additional line for the struct field. For the interface, there are 3 lines for the method, 1 line for an interface assertion, and blank lines around those, so 7 lines. Of course, the two are not interchangeable: the interface is strictly more powerful and can be used when one does not want to define operation from a struct field.

With respect to reflection, PD already uses dynamic reflection any time it constructs a JSON response. Such is the way of go. So the choice is not to use reflection or not to, but instead to choose what constitutes a good usage. Reflection will always have performance concerns, but I assume that at the point we are dealing with translating error data into client responses performance is usually a non-issue and instead we care mostly about correctness. I think this usage of reflection is very simple and straightforward from the perspective of having a reliable and easy to understand implementation, and probably much more so than any other current usage of reflection such as JSON generation.

I did have two concerns with the implementation:
* With an interface one can assert that it is implemented. However, with structural reflection is possible to misspell "Operation" and not realize it.
* structural reflection does not seem common in go so this might seem odd to go programmers

There are other possible approaches to designing the API, so I am open to alternative suggestions. It is possible to try to build convenience on top of `OpErrCode`, but this leads to requiring wrappers for error data construction. However, I did just now think of an approach of embedding `type Op struct { Op: string }`. The implementation is here, let me know what you think. There is some additional minorly annoying boilerplate, but users can still take the same convenient approach as before: https://github.com/pingcap/pd/compare/gregwebs/error-code-operation...gregwebs/error-code-operation-no-reflection?expand=1

Reply:
@gregwebs 

I think the interface is better than dynamic reflection. Mostly, I prefer designing for interface :-)

Reply:
Okay, I made it possible to do the embed or to wrap rather than to reflect, PTAL. I think these approaches offer enough convenience and they avoid the possibility of mistake by typo.